sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_azureappconfigurationprovider.py

```diff
@@ -259,11 +256,151 @@ class AzureAppConfigurationProvider(Mapping[str, Union[str, JSON]]):
+
+        if self._refresh_options is not None:
+            self._next_refresh_time = datetime.now() + timedelta(
+                seconds=self._refresh_options._refresh_interval
+            )
+
+        trim_prefixes: List[str] = kwargs.pop("trim_prefixes", [])
+        self._trim_prefixes = sorted(trim_prefixes, key=len, reverse=True)
+        self._attempts = 1
+        self._min_backoff: Optional[int] = kwargs.pop("min_backoff", 30)
+        self._max_backoff: Optional[int] = kwargs.pop("min_backoff", 600)
+
+    def refresh(self) -> None:
+        if (
+            self._refresh_options is None
+            or len(self._refresh_options.get_refresh_registrations()) == 0
+        ):
+            return
+
+        if datetime.now() < self._next_refresh_time:
+            return
+
+        refresh_all = False
+
+        for registration in self._refresh_options.get_refresh_registrations():
+            if registration.refresh_all == True:
+                updated_sentinel = self._client.get_configuration_setting(
+                    key=registration.key_filter, label=registration.label_filter
+                )
+                if registration.etag == updated_sentinel.etag:
+                    continue
+                else:
+                    refresh_all = True
+                    break
+
+        if refresh_all:
+            self._load_all()
+            self._next_refresh_time = datetime.now() + timedelta(
+                seconds=self._refresh_options._refresh_interval
+            )
+            return
+
+        # Only update individual keys if the refresh_all didn't trigger
+        updated_dict = self._dict.copy()
+        updated_registrations = self._refresh_options.get_refresh_registrations().copy()
+        
+        try:
+            for registration in updated_registrations:
+                if registration.refresh_all == False:
+                    updated_sentinel = self._client.get_configuration_setting(
+                        key=registration.key_filter, label=registration.label_filter
+                    )
+                    if registration.etag == updated_sentinel.etag:
+                        continue
+                    else:
+                        registration.etag = updated_sentinel.etag
+                        updated_dict[
+                            self._proccess_key_name(updated_sentinel)
+                        ] = self._proccess_key_value(updated_sentinel)
+            
+            self._dict = updated_dict
+            self._refresh_options.set_refresh_registrations(updated_registrations)
+            self._next_refresh_time = datetime.now() + timedelta(
+                seconds=self._refresh_options._refresh_interval
+            )
+            self._attempts = 1
+        except:
+            # Refresh All or None, any failure will trigger a backoff
+            self._next_refresh_time = datetime.now() + timedelta(microseconds=self._calculate_backoff())
+            self._attempts += 1
```

Comment:
It seems that we never check that max_attempts has been hit, so we will keep retrying infinitely?